### PR TITLE
Add Parser for Simulink Check tool

### PR DESCRIPTION
**Pull request by evahabeeballah**
This pull request is to include a parser for .html report files generated from Simulink Check tool by Mathworks.
Pull request for warnings-ng-plugin link will be done after this request is approved.
Link for Simulink Check tool: https://de.mathworks.com/products/simulink-check.html

The following files have been added:
SimulinkCheckParser.java : Parser file, using correctly parse and go through  html files.
SimulinkCheckDescriptor.java: Descriptor for Simulink Check parser.
SimulinkCheckParserTest.java: Tester for Simulink Check parser.
Simulink.html: Sample html file used for tester.

The following files were edited:
SUPPORTED-FORMATS.md
pom.xml
ParserRegistry.java 

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+ * Used for .html report files.
+ *
+ * @author Eva Habeeb
+ */
+public class SimulinkCheckParser extends IssuePropertiesParser {
+
+    private static final long serialVersionUID = -8099258658775128275L;
+
+    @Override
+    @SuppressWarnings({"PMD.NcssCount", "PMD.CyclomaticComplexity", "PMD.CognitiveComplexity", "PMD.NPathComplexity" })
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java:27]: Is this warning related to the code in line 28-53? There is no single `if` contained.

File: src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    public void parseIssue(final Report report, final Document document, final IssueBuilder issueBuilder, final String system, final String check) {
+        Elements heading;
+        String issueTxt;
+        String headingElement;
+        String[] headingSplit;
+        int strLen;
+        String module;
+
+        Elements issueElements = document.select(check);
+        for (Element w : issueElements) {
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java:73]: Are all other those HTML methods `null` safe? So what happens if not? IF the HTML document will be changed slightly then everything breaks with NPE.

File: src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+            heading = w.select("span.CheckHeading");
+            issueTxt = heading.text();
+            setSeverity(check, issueBuilder);
+
+            headingElement = heading.first().id();
+            headingSplit = headingElement.split("\\.");
+            strLen = headingSplit.length;
+
+            if (issueTxt.contains("SW")) {
+                String[] fileName = issueTxt.split(":", 2);
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java:83]: What happens, if not present? Then the index is not accessible

File: src/main/java/edu/hm/hafner/analysis/registry/SimulinkCheckDescriptor.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    @Override
+    public String getUrl() {
+        return "https://www.mathworks.com/products/simulink-check.html";
+    }
+
+    @Override
+    public String getHelp() {
+        return "Reads and Parses HTML reports of Simulink Check Tool by MathWorks. "
+                + "Report can be generated with command: "
+                + "ModelAdvisor.summaryReport(ModelAdvisor.run(<SYSTEMS>,<CONFIG>,<FILENAME>,<ARGS>))";
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/registry/SimulinkCheckDescriptor.java:35]: Wrap command in code tags.

**Review by uhafner — CHANGES_REQUESTED** [review]

File: src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+    /**
+     * Handles parsing.
+     * @param report the report object
+     * @param document parsed html document
+     * @param issueBuilder issueBuilder object
+     * @param system name of the system
+     * @param check category of the check
+     *
+     */
```
[the remark this answers is not in the document]
Comment (evahabeeballah) [reply src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java]: Fixed.

File: src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    public void parseIssue(final Report report, final Document document, final IssueBuilder issueBuilder, final String system, final String check) {
+        Elements heading;
+        String issueTxt;
+        String headingElement;
+        String[] headingSplit;
+        int strLen;
+        String module;
```
[the remark this answers is not in the document]
Comment (evahabeeballah) [reply src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java]: Fixed.

File: src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+            report.add(issueBuilder.build());
+        }
+    }
+
+    /**
+     * Handles setting severity of issues.
+     * @param check category of the check
+     * @param issueBuilder issueBuilder object
+     */
+    public void setSeverity(final String check, final IssueBuilder issueBuilder) {
```
[the remark this answers is not in the document]
Comment (evahabeeballah) [reply src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java]: Fixed.

File: src/test/java/edu/hm/hafner/analysis/parser/SimulinkCheckParserTest.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+import edu.hm.hafner.analysis.AbstractParserTest;
+import edu.hm.hafner.analysis.Report;
+import edu.hm.hafner.analysis.Severity;
+import edu.hm.hafner.analysis.assertions.SoftAssertions;
+
+class SimulinkCheckParserTest extends AbstractParserTest {
+
+    SimulinkCheckParserTest() {
+        super("Simulink.html");
```
[the remark this answers is not in the document]
Comment (evahabeeballah) [reply src/test/java/edu/hm/hafner/analysis/parser/SimulinkCheckParserTest.java:11]: Fixed.

File: src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java (hunk 1 above)
Comment (evahabeeballah) [reply src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java]: Fixed, sorry it was left there from the old version of the code.

File: src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java (hunk 2 above)
Comment (evahabeeballah) [reply src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java:58]: Yes the html methods used here are null safe, they return empty 'Element' objects if any of the check categories are not found in the report. All the HTML reports for simulink check have the same structure, the difference is that some reports might not have all the check categories, which is handled by the 'select' function.

File: src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java (hunk 3 above)
Comment (evahabeeballah) [reply src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java:74]: This case is added for a specific scenario where the issue text will for sure have the value SW0 and ':', otherwise it will be handled by the other else statements. I have updated the condition to "if (issueTxt.contains("SW0"))" to make sure it is the correct case.

File: src/main/java/edu/hm/hafner/analysis/registry/SimulinkCheckDescriptor.java (hunk 4 above)
Comment (evahabeeballah) [reply src/main/java/edu/hm/hafner/analysis/registry/SimulinkCheckDescriptor.java]: Fixed.

File: src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+ * @author Eva Habeeb
+ */
+public class SimulinkCheckParser extends IssuePropertiesParser {
+
+    private static final long serialVersionUID = -8099258658775128275L;
+
+    @Override
+    public Report parse(final ReaderFactory readerFactory) throws ParsingException {
+        String name = readerFactory.getFileName();
+        File file = new File(name);
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java:29]: Use the reader factory to read the file. The filename might not be set correctly.

File: src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+            throw new ParsingException(e);
+        }
+    }
+
+    private void parseIssue(final Report report, final Document document, final IssueBuilder issueBuilder, final String system, final String check) {
+
+        Elements issueElements = document.select(check);
+        for (Element w : issueElements) {
+            Elements heading;
+            heading = w.select("span.CheckHeading");
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java:59]: Initialize directly

File: src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+            String issueTxt;
+            issueTxt = heading.text();
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java:61]: same here and below

File: src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+                issueBuilder.setAdditionalProperties(w.parent().id());
+            }
+            issueBuilder.setFileName(system);
+            report.add(issueBuilder.build());
+        }
+    }
+
+    private void setSeverity(final String check, final IssueBuilder issueBuilder) {
+
+        if ("div.WarningCheck".equals(check)) {
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java:99]: The String is already used above: use a constant

**Review by uhafner — CHANGES_REQUESTED** [review]

File: src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java (hunk 9 above)
Comment (evahabeeballah) [reply src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java:29]: There is no function in redearFactrory that reads the file and returns object with type 'File' which is required for the Jsoup.parse function in line 33.

Comment (uhafner) [reply src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java:29]: Why can't you use one of the other methods that actually work on a `Reader` or `InputStream`?

Comment (evahabeeballah) [reply src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java:29]: Which methods do you mean, the methods in the ReaderFactory class ?

Comment (uhafner) [reply src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java:29]: Jsoup has also methods that accept a `Reader`, and the factory can create a `Reader` as well.

Comment (evahabeeballah) [reply src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java:29]: I tried to use parse method from Jsoup that accepts 'InputStream' : parse(InputStream in, String charsetName, String baseUri), but the 'shouldHandleEmptyFile' test fails. 
Jsoup doesnt have a parse method that accepts 'Reader' type so I had to convert from 'Reader' to 'InputStream'. (code snapshot attached) 
image

Comment (uhafner) [reply src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java:29]: This looks good! When the test fails then you need to fix the code: empty files should either throw a `ParsingException` or an empty list of issues.

Is the encoding always UTF-8? If not: the reader factory already has the correct encoding stored.

Comment (uhafner) [reply src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java:29]: BTW: does JSOUP load included elements of the HTML document from the internet? How does JSOUP handle JS in the HTML document? Do we need to add XSS protection as we are doing for XML parsers?

Comment (evahabeeballah) [reply src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java]: Added a ParsingException to handle empty files. As for the encoding, Ive check for all the sample files I have and it is always UTF-8, but just incase I changed it to use getCharset() from ReaderFactory.

Comment (evahabeeballah) [reply src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java]: image

According to the documentation of JSOUP, to prevent XSS attacks I can use their Cleaner class which is:
 "a safelist based HTML cleaner. Use to ensure that end-user provided HTML contains only the elements and attributes that you are expecting; no junk, and no cross-site scripting attacks!
Safe-lists define what HTML (elements and attributes) to allow through the cleaner. Everything else is removed"

File: src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java (hunk 10 above)
Comment (evahabeeballah) [reply src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java]: Fixed

File: src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java (hunk 11 above)
Comment (evahabeeballah) [reply src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java]: Fixed

File: src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java (hunk 12 above)
Comment (evahabeeballah) [reply src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java:116]: Fixed

File: src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java (hunk 9 above)
Comment (uhafner) [reply src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java]: This looks like a solution to get the HTML clean. What I mean: does JSOUP load content from the internet when placed in the HTML? If yet it might make sense to use the cleaner before the parsing. But if JSOUP does not load anything else (only the provided document) then we are on the safe side.

Comment (evahabeeballah) [reply src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java]: I am not able to find any information on whether JSOUP loads the internet content inside the HTML while using the parse function. The documentation states that the parse function 'reads input stream and parses it into Document'.

image

I believe it wont be possible to use cleaner before parsing because the cleaner expects an 'org.jsoup.nodes.Document'.

Comment (uhafner) [reply src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java]: Ok, thanks for checking!

File: src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+    public Report parse(final ReaderFactory readerFactory) throws ParsingException {
+        try (IssueBuilder issueBuilder = new IssueBuilder()) {
+            Reader reader = readerFactory.create();
+            InputStream targetStream = new ReaderInputStream(reader, readerFactory.getCharset());
+
+            //Create a safe-list with allowed tags and attributes to prevent XSS attacks
+            Safelist safelist = new Safelist();
+            safelist = safelist.basic().addTags("div");
+            safelist = safelist.addAttributes("div", "class", "id");
+            safelist = safelist.addAttributes("span", "class", "id");
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java:48]: Sorry, but this is not needed anymore *after* Jsoup parsed the document. We simply can delete that part.

File: src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+    static final String NOTRUN = "div.NotRunCheck";
+    static final String INCOMPLETE = "div.IncompleteCheck";
+
+    @SuppressWarnings({"PMD.AvoidCatchingNPE", "PMD.CloseResource"})
+    @SuppressFBWarnings(value = "DCN_NULLPOINTER_EXCEPTION", justification = "throws parsing exception")
+    @Override
+    public Report parse(final ReaderFactory readerFactory) throws ParsingException {
+        try (IssueBuilder issueBuilder = new IssueBuilder()) {
+            Reader reader = readerFactory.create();
+            InputStream targetStream = new ReaderInputStream(reader, readerFactory.getCharset());
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java:42]: These need to be part of the try (...) Then it gets automatically closed.

File: src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+public class SimulinkCheckParser extends IssuePropertiesParser {
+
+    private static final long serialVersionUID = -8099258658775128275L;
+    static final String WARNING = "div.WarningCheck";
+    static final String FAILED = "div.FailedCheck";
+    static final String NOTRUN = "div.NotRunCheck";
+    static final String INCOMPLETE = "div.IncompleteCheck";
+
+    @SuppressWarnings({"PMD.AvoidCatchingNPE", "PMD.CloseResource"})
+    @SuppressFBWarnings(value = "DCN_NULLPOINTER_EXCEPTION", justification = "throws parsing exception")
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java:37]: This does not make sense

File: src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+            //Create a safe-list with allowed tags and attributes to prevent XSS attacks
+            Safelist safelist = new Safelist();
+            safelist = safelist.basic().addTags("div");
+            safelist = safelist.addAttributes("div", "class", "id");
+            safelist = safelist.addAttributes("span", "class", "id");
+
+            //Parse html document using Jsoup
+            Document dirtyDoc = Jsoup.parse(targetStream, String.valueOf(readerFactory.getCharset()), "");
+            reader.close();
+            targetStream.close();
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java:53]: This is part of auto-closable

File: src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+            //Select Failed items from the whole html document
+            parseIssue(report, document, issueBuilder, system, FAILED);
+            //Select Not Run items from the whole html document
+            parseIssue(report, document, issueBuilder, system, NOTRUN);
+            //Select Incomplete items from the whole html document
+            parseIssue(report, document, issueBuilder, system, INCOMPLETE);
+
+            return report;
+        }
+        catch (IOException | NullPointerException e) {
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java:72]: There should no need to catch a NPE in any code!

File: SUPPORTED-FORMATS.md [hunk 18]
```diff
@@ -1804,6 +1804,27 @@ analyze - iccxxxxcompiler_opts cstat2.c</pre></code>For details check the IAR C-
                 -
             </td>
         </tr>
+        <tr>
+            <td>
+                SimulinkCheckParse
```
Comment (uhafner) [inline SUPPORTED-FORMATS.md:1809]: Re-Generate

**Review by uhafner — CHANGES_REQUESTED** [review]

File: src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java (hunk 3 above)
Comment (uhafner) [reply src/main/java/edu/hm/hafner/analysis/parser/SimulinkCheckParser.java:92]: But when they decide to use a - in a future version of the tool then your code breaks with an `IndexOutOfBoundsException`.
